src/ast/parse.rs

```diff
@@ -94,801 +92,785 @@ impl Display for Span {
-        let (x, span, next) = parser(ctx)?;
-        Ok((f(x, span.clone()), span, next))
+        error!(self.span(), "expected {e}")
     }
-}
 
-fn sequence<T, U>(a: impl Parser<T>, b: impl Parser<U>) -> impl Parser<(T, U)> {
-    move |ctx| {
-        let (x, lo, next) = a(ctx)?;
-        let (y, hi, next) = b(&next)?;
-        Ok(((x, y), Span(lo.0, lo.1, hi.2), next))
-    }
-}
-
-macro_rules! sequences {
-    ( $x:expr ) => { $x };
-    ( $x:expr , $( $xs:expr ),+ $(,)? ) => {
-        sequence( $x, sequences!( $( $xs ),+ ) )
-    };
-}
-
-fn sequence3<T, U, V>(
-    a: impl Parser<T>,
-    b: impl Parser<U>,
-    c: impl Parser<V>,
-) -> impl Parser<(T, U, V)> {
-    map(sequences!(a, b, c), |(a, (b, c)), _| (a, b, c))
-}
-
-fn sequence4<T, U, V, W>(
-    a: impl Parser<T>,
-    b: impl Parser<U>,
-    c: impl Parser<V>,
-    d: impl Parser<W>,
-) -> impl Parser<(T, U, V, W)> {
-    map(sequences!(a, b, c, d), |(a, (b, (c, d))), _| (a, b, c, d))
-}
-
-fn option<T>(parser: impl Parser<T>) -> impl Parser<Option<T>> {
-    move |ctx| match parser(ctx) {
-        Ok((x, span, next)) => Ok((Some(x), span, next)),
-        Err(_) => Ok((None, ctx.span(), ctx.clone())),
+    fn expect_call(&self, e: &'static str) -> Result<(Symbol, &[Sexp], Span), ParseError> {
+        if let Sexp::List(sexps, span) = self {
+            if let [Sexp::Atom(func, _), args @ ..] = sexps.as_slice() {
+                return Ok((*func, args, span.clone()));
+            }
+        }
+        error!(self.span(), "expected {e}")
     }
 }
 
-fn parens<T>(f: impl Parser<T>) -> impl Parser<T> {
-    map(
-        choice(
-            sequence3(text("["), f.clone(), text("]")),
-            sequence3(text("("), f, text(")")),
-        ),
-        |((), x, ()), _| x,
-    )
-}
-
-fn program(ctx: &Context) -> Res<Vec<Command>> {
-    repeat_until(command, |ctx| ctx.index == ctx.source.contents.len())(ctx)
-}
-
-fn rec_datatype(ctx: &Context) -> Res<(Span, Symbol, Subdatatypes)> {
-    choice(
-        map(
-            parens(sequence3(
-                text("sort"),
-                ident,
-                parens(sequence(ident, repeat_until_end_paren(expr))),
-            )),
-            |((), name, (head, exprs)), span| (span, name, Subdatatypes::NewSort(head, exprs)),
-        ),
-        map(
-            parens(sequence(ident, repeat_until_end_paren(variant))),
-            |(name, variants), span| (span, name, Subdatatypes::Variants(variants)),
-        ),
-    )(ctx)
-}
-
-fn list<T>(f: impl Parser<T>) -> impl Parser<Vec<T>> {
-    parens(repeat_until_end_paren(f))
+// helper for mapping a function that returns `Result`
+fn map_fallible<T>(
+    slice: &[Sexp],
+    func: impl Fn(&Sexp) -> Result<T, ParseError>,
+) -> Result<Vec<T>, ParseError> {
+    slice.iter().map(func).collect::<Result<_, _>>()
 }
 
-fn snd<T, U>(x: Option<(T, U)>, _span: Span) -> Option<U> {
-    x.map(|(_, x)| x)
-}
+fn command(sexp: &Sexp) -> Result<Command, ParseError> {
```

Comment:
Make this function an implementation of some trait, so that we can support user-defined keywords?

Reply:
This is future work